interfaces/builtin/nvidia_support.go

```diff
@@ -0,0 +1,82 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const nvidiaSupportSummary = `allows creating static and dynamic NVIDIA chrdev`
+
+const nvidiaSupportBaseDeclarationPlugs = `
+  nvidia-support:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const nvidiaSupportBaseDeclarationSlots = `
+  nvidia-support:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const nvidiaSupportConnectedPlugAppArmor = `
+# This is inverse of
+# https://forum.snapcraft.io/t/call-for-testing-chromium-62-0-3202-62/2569/46
+# As nvidia-assemble snap needs to create the static & dynamic MAJOR
+# chrdevs for all other snaps to have access to. Specifically
+# /dev/nvidiactl /dev/nvidia-uvm
+/{,usr/}bin/mknod ixr,
+allow capability mknod,
+
+/dev/nvidia[0-9]* rw,
+/dev/nvidiactl rw,
+/dev/nvidia-uvm rw,
+/dev/nvidia-uvm-tools rw,
+/dev/nvidia-modeset rw,
+
+# To read dynamically allocated MAJOR for nvidia-uvm
+@{PROC}/devices r,
+`
+
+const nvidiaSupportConnectedPlugSecComp = `
+# This is inverse of
+# https://forum.snapcraft.io/t/call-for-testing-chromium-62-0-3202-62/2569/46
+# As nvidia-assemble snap needs to create the static & dynamic MAJOR
+# chrdevs for all other snaps to have access to. Specifically
+# /dev/nvidiactl /dev/nvidia-uvm
+mknod - |S_IFCHR -
+mknodat - - |S_IFCHR -
+`
+
+type nvidiaSupportInterface struct {
+	commonInterface
+}
+
+func init() {
+	registerIface(&nvidiaSupportInterface{commonInterface: commonInterface{
+		name:                  "nvidia-support",
```

Comment:
nvidia-nodes-support or nvidia-device-nodes-support would seem a bit more explicit about what this is about, usually in foo-support foo is  a snap or service, calling this nvidia-assemble-support seem too specific, but otoh nvidia-support seems to be about supporting a range of potential NVIDIA bits which doesn't seem to be the case

Reply:
There are more things that nvidia could do here - there are extra daemons to manager suspend / resume / power control, and additional device node types (dynamically created and delegated) that could be done, plus this is only for a single card support, whereas multi-card support will need write access to /dev/nvidia[0-9]* and similar changes. Thus yes, this is very bare bones nvidia support, for a single GPU, GL or CUDA usecase. A lot more things will need to be opened up to allow advance nvidia usage. THere is a lot more things one can do with nvidia-caps, nvidia-nvswitch, nvidia-nvlink, etc.


For just the device nodes, it would be nice for this to exist:

``` yaml
plugs:
  nvidia:
    interface: kernel-device-node
    device-nodes:
    - path: /dev/nvidiactl # path to the new device node
      type: block|character|fifo # type of the new device node
      major: 195 # statically allocated major number
      minor: 0 # staticially allocated minor number, or omitted for any, possibly want a range here ie. 0-255
      mode: 666 # file mode of the new device node
      user: root # user owning the new device node
      group: root # group owning the new device node
    - path: /dev/nvidia-uvm
      type: block|character|fifo
      module: nvdia-uvm # this means major number is dynamically allocated, and must be looked up from /proc/devices at creation time, if we can't enforce it, it should be super-privileged `any`
      minor: 255
      mode: 666
      owner: root
      group: root
```
with a matching `snapctl mknod` command to do that would be nice and generic. The overall permissions would allow creating such files, and execute mknodat. I don't know how dynamically looked up major can be enforced with like apparmor or seccomp.

in practice, all open source drivers do that automatically in kernel space, and existing unconfined udev rules apply mode/owner/group changes too. The reason why nvidia driver does not do this, is because the kernel APIs to do this are exported to GPL modules only. That's why all nvidia tooling constantly tries to delete/recreate these nodes in userspace (i.e. all the cuda, nvidia gl libraries, chrome). Thus a generic interface interface is of limited value for the opensource drivers and harder for me to code. Hence this specific interface, is compact, and straightforward to review. Rather than accepting a general purpose interface, and relying on store reviews to ensure that nobody but nvidia-assemble snap use this.

Reply:
if we plan to expand the usage of this to also services/daemons and as we might not want to create many single-use interfaces which is going to be confusing/polluting the interface namespace, the best approach is probably to generalize the summary of this interface, maybe call it nvidia-drivers-support and use/be prepared to  use a "flavor" mechanism like the kubernetes-support interface does (see kubernetes_support.go)

Reply:
Yes, something like kubernetes-support may come up in the future. If we package other components of userspace driver support, or if we require conflicting/alternative ones (with/without persistence, with/without desktop support, with/without cuda support). And yes this is to do with the driver itself, like kubernetes daemon itself. Applications that want to use opengl or cuda acceleration, only need `opengl` interface.